Feature: 14621 Tracktiles (Part 1)

ralflang opened the pull request:
# Base Feature for Track Tiles

Implements first parts for issue 14621 - Track Tiles
Supersedes PR #14634 

Redone implementation. The first version #14634 was not supposed to show up in jmri/jmri but in ralflang/jmri and contained too much noise.

Includes two commits by @bobjacobsen 

## Functionality implemented

- Tracktile catalogs for PIKO A and Märklin C track families, 50 pieces total
- Tracktile XSD and XSLT
- Some documentation on turnouts
- Tracktile catalog viewer (table with filters)
- LayoutEditor UI to select tracktiles when placing TrackSegments
- LayoutEditor UI flow to place a tracktile with correct length, radius and orientation wrt a previous tracktile (see below)
- LayoutEditor Edit Track Segment dialog shows if it is a tile or not and displays the orientation at both ends of a track segment, correctly calculating for straights and simple curves. I don't have any times with bezier forms yet.
- Changes to the XSD and loading code for Layouts to correctly save and load TrackTiles.
- Some foundation for supporting turnouts and slips in the next part
- Some unit tests for the math

## How to use this (partial) feature

- Place one PositionalPoint (anchor or other)
- Select Track Segment
- Select a tile from pulldown menus vendor, family, tile

### Initial tile placing

- Shift-Click on the anchor point and drag in the direction of initial orientation. Release
- The tile will be placed with the correct length. If it is a straight, it will have the direction of the cursor drag. If it is a curve, it will start in the direction of the cursor drag and end at the tile's builtin arc.
- At the end of the tile an anchor point will be created and attached. If a PositionalPoint point already exists, it will be used and attached. Attaching to existing connectors on turnouts et al is not yet supported.

### Consecutive tiles

- Shift-Click on the initial tile's anchor point. Both ends work.
- Drag in any direction, does not matter. Release
- The tile will begin at the position and orientation of the previous tile's end and will snap into the correct position. The cursor drag direction does not matter.

### Non-Tiles

- Create a target anchor point
- Select "track segment". It should have reset to "not a tile".
- Shift-click a free anchor next to a tile and connect to the target anchor. Release
- The non-tile will be created as usual. No restrictions apply.

## Deforming tiles after the fact

In the current implementation there is no protection of tiles against the user visually distorting and editing them.
There is a large gap between a rigid "physical" model and what JMRI users according to the mailing list are used to do.

I'd like to implement a logically connected board "at scale" but this is more involved and distracting from current goals.

## Procedure

After release of the current JMRI version I'd like to get this foundation part through review with all the necessary critique.
In parallel I am working on turnout support but this should be a separate PR. The current one is large enough.

If reviewers feel I should split down further, just give me a hint.

## Testing

This ran on my machine including new unit tests. I will follow up on anything the CI comes up with.

bobjacobsen committed:
updated schema and validating definition files

bobjacobsen committed:
add tests of schema and definition files

ralflang committed:
feat(xmltracktiles): Tracktiles XSD, Piko A catalog, Marklin C catalog

ralflang committed:
feat(xmltracktiles): XSLT, types

ralflang committed:
feat(LayoutTileMath): Base curve and straight calculation

ralflang committed:
feat(tracktileviewer): Implement a TrackTile Table Viewer with filters in the Tables Menu

ralflang committed:
style: whitespace and line breaks

ralflang committed:
ci: Add tag 'tracktiles' to Bob's TrackTiles Schema test for running this specific subset

ralflang committed:
fix(tracktiles): Marklin crossover must be xover

ralflang committed:
docs: LayoutEditor turnout path logic

ralflang committed:
feat(tracktiles): LayoutEditor Toolbar variant tile selection

Horizontal, vertical and floating toolbars now allow to select
vendor/family/tile based on the editor tool selected.

Additional radio buttons for left/right curves and for the different
connectors of a turnout show up.

ralflang committed:
feat(xmltracktiles): Add missing route attributes on some turnouts.

feat(tracktiles): Upgrade XML format to make single path tiles more like other tiles

fix(xmltracktiles): Correctly validate path AB

fix: xmltracktiles straight ab path

ralflang committed:
feat: Add interface for LayoutTrack elements which support tiles

ralflang committed:
LayoutTileMath is all static methods

ralflang committed:
feat: Add TrackTile support interface to Turnout and TrackSegment

ralflang committed:
feat: Improved LayoutTileMathTest.java

ralflang committed:
style: WS and formatting

ralflang committed:
feat: getSelectedTrackTile now returns NotATile instead of null

ralflang committed:
feat: Add TrackTile or NotATile when adding TrackSegmentViews

ralflang committed:
feat: getTileOrientationAt for straight and circle segments

ci: Test for TrackTileSegmentView.getTileOrientationAt
ci: Add inverse direction tests to LayoutTileMathTest.java

ralflang committed:
feat(TrackSegmentEditor): Display orientation at endpoints of straight or curve

ralflang committed:
feat: Preliminary implementation of tile mode for straight and circle TrackSegments

fix: Initial curved tile placing is now correct.
fix: Get orientation calculation right for consecutive curves or straights.

ralflang committed:
feat: Helper method to normalize angles to valid circle degrees

ralflang committed:
fix: Demote tile flow logging to debug level

fix: Remove system.out.println debugging from TrackTilePath

ralflang committed:
fix: Update XML Schema (editors.xsd) and XSLT  (panelfile.xsl) for tile support

ralflang committed:
feat: XML Persistence load/save tile association

Implemented for LevelXing, Slip, Turnout and TrackSegment
PositionablePoint don't need them. (TODO: How to handle articulated
tiles with endpoints and track elements)
I won't deal with turntables anytime soon.

ralflang committed:
feat: Display tile information in Edit Track Segment dialog (or not a tile info)

ralflang committed:
feat: Implement stub DefaultTrackTileManagerXml - at the moment readonly

ralflang committed:
fix: Static Analyzer complaints

fix: umlaut in help files

dsand47 commented:
I recommend that there be a Layout Editor Option to enable Tiles.  This would primarily be used control toolbar clutter.  It may have other capabilities later.  The attribute would only be included in the xml when true.

For the floating tool box, it would help to stack the Vendor/Family/Tile inputs.  The window resizing is disruptive.

The Edit detail for a standard track segment is confusing.

<img width="340" height="344" alt="Screenshot 2025-12-04 at 2 39 39 PM" src="https://github.com/user-attachments/assets/3e2a70e9-7dfe-4fac-ae62-a651509efa03" />

ralflang commented:
I have fixed all the static linter stuff. I'd need some guidance on this unit test result:

```
Error:    CheckHelpFilesForUTF8Test.tearDown:132 Unexpected ERROR or higher
 messages emitted: "Num errors: 2" ==> expected: <false> but was: <true>
```

ralflang committed:
style: Undo multiple file reformatting

style: Undo non-functional changes to ListedTableFrame.java

style: undo non-functional changes to LayoutTurnoutView.java

style: undo ws changes to ToolsMenu.java

style: Undo LayoutEditor.java nonfunctional changes

ralflang committed:
doc: Update turnout_paths.md according to review feedback

dsand47 commented on java/src/jmri/jmrit/display/layoutEditor/turnout_paths.md:
````diff
@@ -0,0 +1,175 @@
+```
+
+**Valid Routes:**
+- **A-C** (ROUTE_AC) - throat to other route (thrown)
````
`one` and `other` is ambiguous.  AC is `left leg` and AB is `right leg`.

ralflang committed:
test: Add required JUnitUtil setUp/tearDown to LayoutTileMathTest

Add @BeforeEach and @AfterEach methods with JUnitUtil.setUp() and
JUnitUtil.tearDown() calls to comply with JMRI architecture rules.
This fixes TestArchitectureTest failure in Static Analysis CI.

ralflang committed:
doc: Fix Wye turnout connection labels in turnout_paths.md

- Swap B and C labels in Wye turnout ASCII diagram to match standard
  JMRI convention where B is upper-left and C is lower-right
- Update route descriptions to use "left leg" and "right leg" terminology
  for clarity instead of "left leg route" and "right leg route"

This addresses documentation consistency issues raised in PR review.
